File: rust/src/run_log.rs
//! Records everything a run attempts, schedules and completes, and writes it to
//! a file when execution finishes.
//!
//! The console narrative already describes the run in full — which services
//! were queued for disabling, which users were created, which updates were
//! applied and which failed — but it scrolls away, and on a competition image
//! there is rarely an opportunity to read it as it goes. Every line the `ui`
//! module prints is therefore mirrored here, with markup stripped and a
//! timestamp attached, and flushed to disk at the end.
//!
//! The buffer is process-global because the `ui` helpers are free functions
//! called from every task; threading a logger through the whole `Task` trait
//! would change every signature to record something none of the tasks
//! themselves care about.

use chrono::Local;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

/// Timestamped lines captured so far, in order.
static ENTRIES: Mutex<Vec<String>> = Mutex::new(Vec::new());

/// Append a line to the run log. Blank lines are dropped to keep the log dense.
pub fn record(text: &str) {
    let text = text.trim_end();
    if text.is_empty() {
        return;
    }
    push(format!("[{}] {}", Local::now().format("%H:%M:%S"), text));
}

/// Append a section heading, mirroring a console rule.
pub fn record_section(title: &str) {
    let title = title.trim();
    if title.is_empty() {
        return;
    }
    push(String::new());
    push(format!("=== {title} ==="));
}

/// Append a line with no timestamp, used for structured blocks at the end.
pub fn record_raw(text: &str) {
    push(text.to_string());
}

fn push(line: String) {
    // A poisoned lock would mean another thread panicked mid-log; the log is
    // diagnostic, so recover the buffer rather than propagating the panic.
    let mut entries = match ENTRIES.lock() {
        Ok(entries) => entries,
        Err(poisoned) => poisoned.into_inner(),
    };
    entries.push(line);
}

/// Everything recorded so far.
/// Record a diagnostic: something the operator does not need to see while the
/// run is happening, but needs afterwards to work out why it did what it did.
///
/// The console narrative reports outcomes - "Failed to remove: CCleaner" - but
/// not the evidence. When the reason comes back empty, which is what happens
/// whenever a tool reports failure only through an exit code, that line degrades
/// to "Failed to remove: CCleaner ()" and there is nothing left to investigate
/// with. Diagnostics carry the evidence: the exact command, its exit code, and
/// what it printed.
///
/// These go to the log only, never to the console. They exist to be read after
/// the fact, and putting them on screen would bury the narrative the operator
/// does have to follow live.
pub fn diagnostic(category: &str, detail: &str) {
    let detail = detail.trim_end();
    if detail.is_empty() {
        return;
    }

    let stamp = chrono::Local::now().format("%H:%M:%S");
    let mut lines = detail.lines();
    if let Some(first) = lines.next() {
        push(format!("[{stamp}] [{category}] {}", first.trim_end()));
    }
    // Indented continuation lines keep a multi-line payload - a captured stderr,
    // say - visibly attached to its own entry rather than reading as separate
    // events.
    for line in lines {
        push(format!("                      {}", line.trim_end()));
    }
}

/// Longest captured output kept per failed command.
const MAX_CAPTURED_OUTPUT: usize = 600;

/// Record the outcome of an external command, with enough to reproduce it.
///
/// Output is captured only on failure, and truncated. A successful command's
/// output is usually large and never interesting; a failure's first few lines
/// are almost always the whole story.
pub fn record_command(
    program: &str,
    arguments: Option<&str>,
    exit_code: Option<i32>,
    output: &str,
    error: Option<&str>,
    elapsed: std::time::Duration,
) {
    let status = match exit_code {
        Some(code) => format!("exit {code}"),
        None => "no exit code (timed out or never ran)".to_string(),
    };

    diagnostic(
        "cmd",
        format!("{program} {}", redact(arguments.unwrap_or_default())).trim_end(),
    );
    diagnostic(
        "cmd",
        &format!("  -> {status} in {:.1}s", elapsed.as_secs_f64()),
    );

    if exit_code == Some(0) {
        return;
    }

    if let Some(err) = error.filter(|e| !e.trim().is_empty()) {
        diagnostic("cmd", &format!("  stderr: {}", truncate(err)));
    }
    if !output.trim().is_empty() {
        diagnostic("cmd", &format!("  stdout: {}", truncate(output)));
    }
}

fn truncate(text: &str) -> String {
    let text = text.trim();
    if text.chars().count() <= MAX_CAPTURED_OUTPUT {
        return text.to_string();
    }
    let kept: String = text.chars().take(MAX_CAPTURED_OUTPUT).collect();
    let remaining = text.chars().count() - MAX_CAPTURED_OUTPUT;
    format!("{kept}... [{remaining} more chars]")
}

/// Blank out plaintext passwords before a command line reaches the log.
///
/// Account passwords are interpolated into `ConvertTo-SecureString` calls. The
/// log does deliberately record each generated password once, where the task
/// announces it - but that is a considered disclosure in one place, not a reason
/// to scatter the same secret through every command echo.
pub fn redact(command_line: &str) -> String {
    let Some(at) = command_line.find("ConvertTo-SecureString") else {
        return command_line.to_string();
    };

    let rest = &command_line[at..];
    let Some(open) = rest.find('\'') else {
        return command_line.to_string();
    };
    let Some(close_rel) = rest[open + 1..].find('\'') else {
        return command_line.to_string();
    };
    let close = open + 1 + close_rel;

    format!(
        "{}{}'***'{}",
        &command_line[..at],
        &rest[..open],
        &rest[close + 1..]
    )
}

pub fn entries() -> Vec<String> {
    match ENTRIES.lock() {
        Ok(entries) => entries.clone(),
        Err(poisoned) => poisoned.into_inner().clone(),
    }
}

/// Discard everything recorded so far. Used by tests to isolate cases.
pub fn clear() {
    match ENTRIES.lock() {
        Ok(mut entries) => entries.clear(),
        Err(poisoned) => poisoned.into_inner().clear(),
    }
}

/// Default log location: the desktop of the user running the tool, alongside
/// the backup folder the prohibited-media task writes.
///
/// The version is part of the file name so logs from different builds are
/// distinguishable at a glance, without opening them.
pub fn default_log_path() -> PathBuf {
    crate::app_config::desktop_dir().join(format!(
        "PinnacleCyPat_RunLog_v{}_{}.txt",
        crate::app_config::VERSION,
        Local::now().format("%Y%m%d_%H%M%S")
    ))
}

/// Write the log to `path`, creating parent directories as needed.
pub fn write_to(path: &Path) -> std::io::Result<()> {
    if let Some(parent) = path.parent() {
        if !parent.as_os_str().is_empty() {
            std::fs::create_dir_all(parent)?;
        }
    }
    let mut body = entries().join("\r\n");
    body.push_str("\r\n");
    std::fs::write(path, body)
}

/// Build the header written at the top of every log.
pub fn header(command_line: &str) -> Vec<String> {
    vec![
        "=".repeat(79),
        "PinnacleCyPat - Run Log".to_string(),
        format!("Version:   {}", crate::app_config::version_string()),
        format!("Started:   {}", Local::now().format("%Y-%m-%d %H:%M:%S")),
        format!("Command:   {command_line}"),
        "=".repeat(79),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;

    // The buffer is global, so these cases must not run concurrently with each
    // other. They are grouped into one test rather than relying on test-thread
    // ordering.
    #[test]
    fn records_strips_blanks_and_writes_a_file() {
        clear();

        record("Disabling service: TlntSvr");
        record("   ");
        record("");
        record_section("Step 2");
        record("Created user: alice");

        let entries = entries();
        assert_eq!(
            entries.iter().filter(|l| l.contains("Disabling service")).count(),
            1
        );
        assert!(
            entries.iter().any(|l| l == "=== Step 2 ==="),
            "section heading missing: {entries:?}"
        );
        assert!(
            entries.iter().any(|l| l.contains("Created user: alice")),
            "entry missing: {entries:?}"
        );
        // Blank input must not produce timestamped empty lines.
        assert!(
            !entries.iter().any(|l| l.ends_with("] ")),
            "blank line recorded: {entries:?}"
        );

        let path = std::env::temp_dir().join(format!("cpa_runlog_{}.txt", std::process::id()));
        write_to(&path).expect("log should write");
        let written = std::fs::read_to_string(&path).unwrap();
        assert!(written.contains("Disabling service: TlntSvr"));
        assert!(written.contains("=== Step 2 ==="));
        let _ = std::fs::remove_file(&path);

        clear();
    }
}

File: src/Core/Utilities/RunLog.cs
using PinnacleCyPat.Core.Models;

namespace PinnacleCyPat.Core.Utilities;

/// <summary>
/// Records everything a run attempts, schedules and completes, and writes it to
/// a file when execution finishes.
/// </summary>
/// <remarks>
/// The console narrative already describes the run in full - which services were
/// queued for disabling, which users were created, which passwords were set -
/// but it scrolls away, and on a competition image there is rarely an
/// opportunity to read it as it goes. Lines are mirrored here with a timestamp
/// and flushed to disk at the end.
/// </remarks>
public static class RunLog
{
    private static readonly List<string> Entries = new();
    private static readonly Lock Gate = new();

    /// <summary>Append a line. Blank lines are dropped to keep the log dense.</summary>
    public static void Record(string text)
    {
        text = text.TrimEnd();
        if (text.Length == 0)
            return;
        Push($"[{DateTime.Now:HH:mm:ss}] {text}");
    }

    /// <summary>Append a section heading, mirroring a console rule.</summary>
    public static void RecordSection(string title)
    {
        title = title.Trim();
        if (title.Length == 0)
            return;
        Push(string.Empty);
        Push($"=== {title} ===");
    }

    /// <summary>Append a line with no timestamp, for structured blocks.</summary>
    public static void RecordRaw(string text) => Push(text);

    /// <summary>
    /// Record a diagnostic: something the operator does not need to see while the
    /// run is happening, but needs afterwards to work out why it did what it did.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The console narrative reports outcomes - "Failed to remove: CCleaner" -
    /// but not the evidence. When the reason string comes back empty, which is
    /// what happens whenever a tool reports failure only through an exit code,
    /// that line degrades to "Failed to remove: CCleaner ()" and there is
    /// nothing left to investigate with. Diagnostics carry the evidence: the
    /// exact command, its exit code, and what it printed.
    /// </para>
    /// <para>
    /// These go to the log only, never to the console. They exist to be read
    /// after the fact, and putting them on screen would bury the narrative that
    /// the operator does have to follow live.
    /// </para>
    /// </remarks>
    public static void Diagnostic(string category, string detail)
    {
        detail = detail.TrimEnd();
        if (detail.Length == 0)
            return;

        // Indented continuation lines keep a multi-line payload - a captured
        // stderr, say - visibly attached to its own entry rather than reading as
        // separate events.
        var lines = detail.Split('\n');
        Push($"[{DateTime.Now:HH:mm:ss}] [{category}] {lines[0].TrimEnd()}");
        foreach (var line in lines.Skip(1))
            Push($"                      {line.TrimEnd()}");
    }

    /// <summary>
    /// Record the outcome of an external command, with enough to reproduce it.
    /// </summary>
    /// <remarks>
    /// Output is captured only on failure, and truncated. A successful command's
    /// output is usually large and never interesting; a failure's first few lines
    /// are almost always the whole story, and an untruncated capture of, say, a
    /// Chocolatey log would bury every other entry in the file.
    /// </remarks>
    public static void RecordCommand(
        string program,
        string? arguments,
        int? exitCode,
        string output,
        string? error,
        TimeSpan elapsed
    )
    {
        var status = exitCode is null
            ? "no exit code (timed out or never ran)"
            : $"exit {exitCode}";
        Diagnostic("cmd", $"{program} {Redact(arguments ?? string.Empty)}".TrimEnd());
        Diagnostic("cmd", $"  -> {status} in {elapsed.TotalSeconds:F1}s");

        if (exitCode == 0)
            return;

        if (!string.IsNullOrWhiteSpace(error))
            Diagnostic("cmd", $"  stderr: {Truncate(error)}");
        if (!string.IsNullOrWhiteSpace(output))
            Diagnostic("cmd", $"  stdout: {Truncate(output)}");
    }

    /// <summary>Longest captured output kept per failed command.</summary>
    private const int MaxCapturedOutput = 2000;

    /// <summary>How much of a long capture to keep from each end.</summary>
    private const int TruncationEdge = MaxCapturedOutput / 2;

    /// <summary>
    /// Shorten a long capture, keeping both ends.
    /// </summary>
    /// <remarks>
    /// Keeping only the head loses the reason. A failed Chocolatey upgrade opens
    /// with pages of package boilerplate and states the actual error - "Error -
    /// hashes do not match" - at the very end, so a head-only cut recorded
    /// everything except the one line worth having.
    /// </remarks>
    private static string Truncate(string text)
    {
        text = text.Trim();
        if (text.Length <= MaxCapturedOutput)
            return text;

        var head = text[..TruncationEdge].TrimEnd();
        var tail = text[^TruncationEdge..].TrimStart();
        var omitted = text.Length - (2 * TruncationEdge);
        return $"{head}\n    ... [{omitted} chars omitted] ...\n    {tail}";
    }

    /// <summary>
    /// Blank out plaintext passwords before a command line reaches the log.
    /// </summary>
    /// <remarks>
    /// Account passwords are interpolated into <c>ConvertTo-SecureString</c>
    /// calls. The run log does deliberately record each generated password once,
    /// where the task announces it - but that is a considered disclosure in one
    /// place, not a reason to scatter the same secret through every command
    /// echo, where it would also survive into any log a competitor shares.
    /// </remarks>
    public static string Redact(string commandLine) =>
        System.Text.RegularExpressions.Regex.Replace(
            commandLine,
            @"(ConvertTo-SecureString\s+)'(?:[^']|'')*'",
            "$1'***'",
            System.Text.RegularExpressions.RegexOptions.IgnoreCase
        );

    private static void Push(string line)
    {
        lock (Gate)
        {
            Entries.Add(line);
        }
    }

    /// <summary>Everything recorded so far.</summary>
    public static List<string> Snapshot()
    {
        lock (Gate)
        {
            return new List<string>(Entries);
        }
    }

    /// <summary>Discard everything recorded so far. Used by tests.</summary>
    public static void Clear()
    {
        lock (Gate)
        {
            Entries.Clear();
        }
    }

    /// <summary>
    /// Default log location: the desktop of the user running the tool.
    /// </summary>
    /// <remarks>
    /// The version is part of the file name so logs from different builds are
    /// distinguishable at a glance, without opening them.
    /// </remarks>
    public static string DefaultLogPath() =>
        Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.Desktop),
            $"PinnacleCyPat_RunLog_v{AppConfig.Version}_{DateTime.Now:yyyyMMdd_HHmmss}.txt"
        );

    /// <summary>Build the header written at the top of every log.</summary>
    public static IEnumerable<string> Header(string commandLine) =>
        new[]
        {
            new string('=', 79),
            "PinnacleCyPat - Run Log",
            $"Version:   {AppConfig.VersionString}",
            $"Started:   {DateTime.Now:yyyy-MM-dd HH:mm:ss}",
            $"Command:   {commandLine}",
            new string('=', 79),
        };

    /// <summary>Append a structured, per-task record.</summary>
    /// <remarks>
    /// The narrative above says what happened as it happened; this block makes
    /// the outcome of each task greppable without reading the whole run.
    /// </remarks>
    public static void AppendResults(IEnumerable<TaskResult> results)
    {
        RecordRaw(string.Empty);
        RecordRaw(new string('=', 79));
        RecordRaw("TASK RESULTS");
        RecordRaw(new string('=', 79));

        foreach (var result in results)
        {
            RecordRaw(string.Empty);
            RecordRaw($"Task:       {result.TaskName}");
            RecordRaw($"Outcome:    {(result.Success ? "SUCCESS" : "FAILED")}");
            RecordRaw($"Verified:   {(result.Verified ? "yes" : "no")}");
            RecordRaw(
                $"Items:      {result.ItemsAttempted} attempted, {result.ItemsSucceeded} succeeded, "
                    + $"{result.ItemsSkipped} skipped"
            );
            RecordRaw($"Confidence: {result.ConfidencePercent}%");
            foreach (var line in (result.Message ?? string.Empty).Split('\n'))
                RecordRaw($"            {line}");
            if (!string.IsNullOrWhiteSpace(result.ErrorDetails))
            {
                RecordRaw("Issues:");
                foreach (var line in result.ErrorDetails.Split('\n'))
                    RecordRaw($"  - {line}");
            }
        }
    }

    /// <summary>
    /// Start mirroring everything Spectre.Console writes into the log.
    /// </summary>
    /// <remarks>
    /// Hooking the console once is far less error-prone than editing every one
    /// of the hundreds of MarkupLine call sites, and it cannot fall out of sync
    /// as tasks change. Spectre renders to an <see cref="IAnsiConsoleOutput"/>;
    /// wrapping that captures the rendered text, from which the markup has
    /// already been resolved, so the log holds exactly what the operator saw.
    /// </remarks>
    public static void AttachToConsole()
    {
        var original = Spectre.Console.AnsiConsole.Console;
        Spectre.Console.AnsiConsole.Console = Spectre.Console.AnsiConsole.Create(
            new Spectre.Console.AnsiConsoleSettings
            {
                Ansi = Spectre.Console.AnsiSupport.Detect,
                ColorSystem = Spectre.Console.ColorSystemSupport.Detect,
                Out = new LoggingConsoleOutput(original.Profile.Out),
            }
        );
    }

    /// <summary>Tees console output into the run log.</summary>
    private sealed class LoggingConsoleOutput : Spectre.Console.IAnsiConsoleOutput
    {
        private readonly Spectre.Console.IAnsiConsoleOutput _inner;
        private readonly System.Text.StringBuilder _pending = new();

        public LoggingConsoleOutput(Spectre.Console.IAnsiConsoleOutput inner) => _inner = inner;

        public TextWriter Writer => new TeeWriter(_inner.Writer, _pending);

        public bool IsTerminal => _inner.IsTerminal;
        public int Width => _inner.Width;
        public int Height => _inner.Height;

        public void SetEncoding(System.Text.Encoding encoding) => _inner.SetEncoding(encoding);

        private sealed class TeeWriter : TextWriter
        {
            private readonly TextWriter _inner;
            private readonly System.Text.StringBuilder _pending;

            public TeeWriter(TextWriter inner, System.Text.StringBuilder pending)
            {
                _inner = inner;
                _pending = pending;
            }

            public override System.Text.Encoding Encoding => _inner.Encoding;

            public override void Write(char value)
            {
                _inner.Write(value);

                // Buffer until a newline so each log entry is a whole line.
                if (value == '\n')
                {
                    Flush();
                    return;
                }
                if (value != '\r')
                    _pending.Append(value);
            }

            public override void Write(string? value)
            {
                if (value == null)
                    return;
                foreach (var c in value)
                    Write(c);
            }

            public override void Flush()
            {
                if (_pending.Length > 0)
                {
                    // Strip ANSI escapes so the log is plain text.
                    var text = System.Text.RegularExpressions.Regex.Replace(
                        _pending.ToString(),
                        "\\[[0-9;]*[A-Za-z]",
                        string.Empty
                    );
                    _pending.Clear();
                    Record(text);
                }
                _inner.Flush();
            }
        }
    }

    /// <summary>Write the log to <paramref name="path"/>.</summary>
    public static async Task WriteToAsync(string path)
    {
        var parent = Path.GetDirectoryName(path);
        if (!string.IsNullOrEmpty(parent))
            Directory.CreateDirectory(parent);

        await File.WriteAllLinesAsync(path, Snapshot());
    }
}
